File: tibiamaps/convert.py
"""Raw minimap export -> PNG + JSON data directory."""
from __future__ import annotations

from pathlib import Path

from . import bounds as bounds_mod
from .markers import save_markers_json, sort_markers
from .sources import RawMinimapSource


def convert_from_minimap(
    input_dir: Path,
    output_dir: Path,
    *,
    include_markers: bool = True,
    markers_only: bool = False,
    floors: set[int] | None = None,
    log=print,
) -> dict:
    source = RawMinimapSource(Path(input_dir))
    b = source.bounds
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    floor_ids = b['floorIDs']
    if floors is not None:
        floor_ids = [fid for fid in floor_ids if int(fid) in floors]

    if not markers_only:
        for fid in floor_ids:
            z = int(fid)
            log(f'Rendering floor {fid}...')
            source.floor_map_image(z).save(output_dir / f'floor-{fid}-map.png')
            source.floor_path_image(z).save(output_dir / f'floor-{fid}-path.png')
        bounds_mod.save_bounds(output_dir / 'bounds.json', b)

    markers = sort_markers(source.markers()) if include_markers else []
    save_markers_json(output_dir / 'markers.json', markers)

    return {
        'bounds': b,
        'floors_rendered': [] if markers_only else floor_ids,
        'marker_count': len(markers),
    }

File: docs/lib/convert.js
import { sortMarkers } from './markers.js';
import { canvasToPngBytes, getFloorMapCanvas, getFloorPathCanvas } from './render.js';

export async function convertFromMinimap(source, { includeMarkers = true, markersOnly = false, floors = null, onProgress } = {}) {
  let floorIds = source.bounds.floorIDs;
  if (floors) floorIds = floorIds.filter((fid) => floors.has(Number(fid)));

  const files = [];
  if (!markersOnly) {
    for (const fid of floorIds) {
      onProgress?.(`Rendering floor ${fid}...`);
      const z = Number(fid);
      const mapCanvas = await getFloorMapCanvas(source, z);
      if (mapCanvas) files.push({ name: `floor-${fid}-map.png`, data: await canvasToPngBytes(mapCanvas) });
      const pathCanvas = await getFloorPathCanvas(source, z);
      if (pathCanvas) files.push({ name: `floor-${fid}-path.png`, data: await canvasToPngBytes(pathCanvas) });
    }
    files.push({ name: 'bounds.json', data: new TextEncoder().encode(JSON.stringify(source.bounds, null, 4)) });
  }

  const markers = includeMarkers ? sortMarkers(await source.markers()) : [];
  files.push({ name: 'markers.json', data: new TextEncoder().encode(JSON.stringify(markers, null, 4)) });

  return {
    files,
    bounds: source.bounds,
    markerCount: markers.length,
    floorsRendered: markersOnly ? [] : floorIds,
  };
}
